### ISSUE TITLE
Timezone support?

### ISSUE THREAD

**Issue by voidale** [vsivsi/meteor-job-collection#104]
Hi, I'm working on scheduling repeat jobs with Later JS, The problem later js doesn't support timezones I'm looking into ways to use later js with specific timezone support using moment timezone it's possible. But can job-collection handle different timezone specific dates? Any precautions I should take? Thanks

**Comment by vsivsi** [comment]
Dates in Jon collection are stored as EJSON dates. Later.js schedules are stored just as they are generated by Later.js, and only Later.js is used to process them.  So whatever method you devise using the moment.js timezome support should work, so long as it produces either valid JS/EJSON Date objects or valid Later.js schedules.

**Comment by voidale** [comment]
Awesome thanks!

**Comment by vsivsi** [comment]
Jobs are promoted from waiting to ready based on a periodic query that always checks an EJSON date. 

Jobs that use later.js schedules use that library to calculate the next Promotion Date when the job is first saved or when a new job is automatically created by a repeating schedule: 
https://github.com/vsivsi/meteor-job-collection/blob/master/src/shared.coffee#L1042

**Comment by mitar** [comment]
I am reopening this because sadly I think this seems to be something which will have to be solved by this package in some way.

There is no simple way currently to make a repeated job at 8 AM in the morning of every day in a timezone of my user, respecting their daylight saving and other changes to timezones. And I have users all across the globe.

Sadly, later.js does not support timezones. And it seems not very maintained anymore.

So I think something should be done here. Maybe we could add timezone handling ourselves, by allowing the user to specify a timezone, maybe as part of the `later` object, or as an additional options, and then we do the manipulation [from this gist](https://gist.github.com/osteenbergen/38259648aa81f6cf27b3) to apply the timezone.

**Comment by mitar** [comment]
Also, the issue is that the version of later provided by this package uses UTC timezone, so one cannot even change the system's timezone to get at least support for one timezone. :-(

**Comment by vsivsi** [comment]
I've had good luck with [momentjs](http://momentjs.com/) and its add-on timezone support (moment timezone)[http://momentjs.com/timezone/]. It sounds like you are mostly interested in allowing users to work in their own timezones, without necessarily caring about how it is represented server-side (where I would still argue UTC makes the most sense). Without too much trouble, I bet you could use momentjs to handle all of the necessary conversions between your UI and Later.js

I'm not inclined to do any work at all on Later.js itself. Feels too much like a "last dev to touch it now owns it" type scenario, and I'm not going there.

**Comment by mitar** [comment]
The gist above does that. What I am saying is that it is just tricky enough to do it properly, that I think it would be great if we could offer it in this package.

**Comment by mitar** [comment]
The issue is that you cannot really convert "at 8 AM" to a timezone in UI. Mapping to timezone value has to be done at every repetition.

**Comment by mitar** [comment]
For example, if you have daylight saving time in your timezone, than half of the year mapping to timezone value of "8 AM" is different than the other half of the year. You cannot change it in advance really.

**Comment by vsivsi** [comment]
Right, I see. I don't really see a straightforward way to fix that without a major overhaul of Later.js, and as I said, I'm not going there. If you come up with a reasonably general solution, I'd be happy to consider a PR, but this isn't something I have the time to work on right now.

**Comment by mitar** [comment]
Should we add another option to `repeat`, like `timezone`, or should we extend the object returned by the later.js?

**Comment by vsivsi** [comment]
My instinct is to say that it should be entirely a later.js thing. The core functionality (e.g. repeat) should be UTC only. Keeping in mind that repeat only knows about fixed delays (e.g. "every 24 hours starting at 8:00am PST on 2/17/2017 (converted to UTC)", and not "every day at 8am local time in Seattle"). Those semantics should remain unchanged. That functionality could have been added to later.js, but my guess is that Later.js avoided that because it is ferociously complicated to get absolutely correct and guarantee that promise. For just one example: the schedule for Daylight Saving Time changes periodically in different locations around the world. How does a running system automatically keep abreast of all of those changes? What happens if it doesn't?  My guess is that later.js punted on this because they just didn't want to take on all of that complexity and ongoing maintenance work. There is no way that a static, 100% bug free library can get this correct, because the ground truth is constantly shifting. 

As I said, I don't want to go anywhere near it.

**Comment by mitar** [comment]
> Later.js avoided that because it is ferociously complicated to get absolutely correct and guarantee that promise

No, they are also OK with a [PR, but they just do not want to have a required dependency on moment-timezone](https://github.com/bunkat/later/issues/95#issuecomment-194481816).

I do not think it is too complicated, there is a [gist for it](https://gist.github.com/osteenbergen/38259648aa81f6cf27b3).

> For just one example: the schedule for Daylight Saving Time changes periodically in different locations around the world. How does a running system automatically keep abreast of all of those changes? 

You use moment-timezone which includes this information. By updating this dependency, you keep it updated. Because "repeatable" jobs are recreated anyway every time you start your app (and old versions are canceled), the only thing you have to do is update dependency and restart your app, and the job with recreated with timestamp mapped with new timezone data.

> As I said, I don't want to go anywhere near it.

Yes, but you also said:

> I'd be happy to consider a PR, but this isn't something I have the time to work on right now.

This is why I am asking what should be the API of this. So I might consider doing it, or somebody else might. So I would like just to clarify the API and other things about that.

So I see two options;

We add a `timezone` option (whatever parsable by moment-timezone) to `repeat`:

```javascript
.repeat({
  schedule: jc.later.parse.text('at 8 AM'),
  timezone: 'America/Los_Angeles'
})
```

Alternatively, we can do:

```javascript
.repeat({
  schedule: _.extend(jc.later.parse.text('at 8 AM'), {
    timezone: 'America/Los_Angeles'
  })
})
```

Then, the second question is how we store this in database. And we have again two options. We add a new field at top level, `timezone`, or `repeatTimezone`. Or we store it inside `repeatWait` field?

**Comment by vsivsi** [comment]
I think the main issue with having a dependence on MomentTZ is that it updates every 30-60 days on average (sometimes more often). And any given update is probably critical to *someone*. If you bake MomentTZ into your package, then you are essentially committing to tracking MomentTZ closely and updating/publishing your own package on the same "heartbeat".

MomentTZ doesn't appear to include any kind of "autoupdate" mechanism for the TZ data files, and even if it did, I don't want JobCollection to require access to the public Internet to work correctly. I also don't want JobCollection to have to publish a new release anytime MomentTZ does... So the two would need to be entirely decoupled.  JobCollection will work with whatever version of MomentTZ you provide, and if you don't provide one, it works as it does today...

But another consideration is that if/when MomentJS does a breaking `1.0` update, then you will also be forced, at that moment to update to accommodate all breaking changes, because it will probably be the only way top continue to receive TZ data updates.

> Because "repeatable" jobs are recreated anyway every time you start your app (and old versions are canceled)...

That is not the default behavior.  Only if you use the `cancelRepeats: true` option to `job.save()`.

Here are my thoughts. Rather than bake-in MomentTZ specific support for this, I'd prefer to see you provide a more general "hook" within the system to modify job run times in some application specified way.  So thinking very generally, an application can put whatever information it wants into the job `data`. Included in that could be the intended timezone information.

Then, the application could provide an "adjustment" function to JC on the server that is invoked with the Job document, any time a job is about to be saved to the server (this would also cover updates, since `.save()` is the update mechanism as well). That way your application controls the TZ library dependency, all code that touches the TZ library, and all metadata required to make it work.  

The change to JobCollection is minimal (adding the "pre-save" hook), and today's default behavior will remain unchanged. This is very similar to the (undocumented) `scrub()` function hook that applications can provide to remove custom server-side job document data that should not be shared with workers.

There is an additional benefit that this function could have other useful applications.

That way @mitar , if you wish, you could separately publish a MomentTZ using `pre-save` function, that would solve this problem for anyone who wants the functionality.

Thoughts?

**Comment by mitar** [comment]
> If you bake MomentTZ into your package, then you are essentially committing to tracking MomentTZ closely and updating/publishing your own package on the same "heartbeat".

Ehm, no? This is why Meteor uses [semvar](http://semver.org/) versions. This package could depend on one version, but user of this package could bump it to any other compatible version. Only if public APIs of MomentTZ would change, then we would have to change stuff. And it seems the release updates to data with **patch** versions. So this is really easy to bump for somebody. They just change `.meteor/versions` file and this is it.

The bigger issue is that there does not seem to be an up-to-date Meteor package for MomentTZ. :-( But one could in a similar way depend on a NPM module.

One more thing. This could also be made as a weak dependency. If user adds in parallel a MomentTZ package (whichever compatible version), then this support for timezone is enabled in this package.

> So the two would need to be entirely decoupled. JobCollection will work with whatever version of MomentTZ you provide, and if you don't provide one, it works as it does today...

Yes, and this is pretty doable.

> Thoughts?

We could also do it with hooks, I just worry that this will make complicated to find all those cases where a hook should be applied. Maybe it could be easier if we just make a hook for custom rescheduling at repeating.

**Comment by satyavh** [comment]
Arghhhh, no timezone support...

I just found out`after` only accepts a date object, and date object always translates back to the servers current timezone messing up my jobs. Example:

```
scheduledAt = "Wed Feb 14 2018 10:10:00 GMT+0100 (CET)"

moment.tz(scheduledAt, "Pacific/Auckland").startOf('day').format()
> "2018-02-14T00:00:00+13:00" 
Awesome!!! But Job Collection doesn't like this date format, so translate to Date Object:

moment.tz(scheduledAt, "Pacific/Auckland").startOf('day').toDate()
> "Tue Feb 13 2018 12:00:00 GMT+0100 (CET)" 
Bad bad bad!!! I'm back where I started :-(
```

And now Laterjs doesn't support timezones either.
What a mess! 

Seems I need to perform some serious package hacking here ;-)

**Comment by mitar** [comment]
That is true. Repeating jobs are not the only jobs where timezone support could be handy. But for other jobs, where you have a constant time points, you can really do preprocessing yourself. I think you are just doing it in a buggy way above.

**Comment by vsivsi** [comment]
Dates in JobCollection are Meteor EJSON dates, and ultimately, on disk, MongoDB BSON dates, which are the only dates you can write a MongoDB query about. There are no time zones in MongoDB, it all bottoms out in UTC, as any sane server-side utility should. 

https://docs.mongodb.com/manual/reference/bson-types/#date

You can correctly provide dates to JobCollection (or any other Meteor/EJSON based application) in any time zone you would like using either the local timezone (via Js Date) or your timezone of choice (Using third party libraries such as the Moment.js Timezone extensions). It is way, way out of scope for a package like JobCollection to provide its own timezone support. Heck, Meteor itself doesn't even touch this problem except through the EJSON date object type (which doesn't do time zone conversions). 

@mitar is correct in pointing out here that it would be a nice enhancement of the functionality of later.js if it worked properly w.r.t. being able schedule repeating events that are TZ/DST aware. If it had that support 2+ years ago, I would have made JobCollection work correctly with it at the time. As it stands right now, later.js appears to be inactively maintained. That sucks for everyone.

I've offered to add hooks to allow applications that need to schedule TZ/DST sensitive jobs to manage that support in whatever way best meets the app's own requirements. If/when the official later.js distribution adds proper timezone support, only then I will look at any changes required in JobCollection to properly support that. In the meantime, I am not going to do separate explicit custom timezone support on JobCollection, nor am I going to build in an explicit separate dependency on a single timezone library, such as MomentTZ. The hooks to handle app dependent run time adjustment are what I am offering today to address this enhancement request. I'm sure you would prefer that I just "fix this" for you, but for all of the reasons I have already articulated, I'm not able to do that right now. 

Please focus your energy on proposing a hook mechanism that will meet the requirements you see for flexibly supporting time zones with a minimum number of changes to JobCollection (or alternatively, work to officially get it into later.js).

**Comment by mitar** [comment]
So, what you think about:

> Maybe it could be easier if we just make a hook for custom rescheduling at repeating.

So I am thinking more of changing this:

```coffee
    # If doc.repeatWait is a later.js object, then don't run before
    # the first valid scheduled time that occurs after doc.after
    if @later? and typeof doc.repeatWait isnt 'number'
      unless next = @later?.schedule(doc.repeatWait).next(1, doc.after)
        console.warn "No valid available later.js times in schedule after #{doc.after}"
        return null
      nextDate = new Date(next)
      unless nextDate <= doc.repeatUntil
        console.warn "No valid available later.js times in schedule before #{doc.repeatUntil}"
        return null
      doc.after = nextDate
    else if not @later? and doc.repeatWait isnt 'number'
      console.warn "Later.js not loaded..."
      return null
```

To this:

```coffee
    if @repeatJob? and @repeatJob doc
      # There is a custom hook to process repeatWait defined, and it returned true
      # to signal that it processed the doc (potentially setting doc.after to a new value).
    # If doc.repeatWait is a later.js object, then don't run before
    # the first valid scheduled time that occurs after doc.after
    else if @later? and typeof doc.repeatWait isnt 'number'
      unless next = @later?.schedule(doc.repeatWait).next(1, doc.after)
        console.warn "No valid available later.js times in schedule after #{doc.after}"
        return null
      nextDate = new Date(next)
      unless nextDate <= doc.repeatUntil
        console.warn "No valid available later.js times in schedule before #{doc.repeatUntil}"
        return null
      doc.after = nextDate
    else if not @later? and doc.repeatWait isnt 'number'
      console.warn "Later.js not loaded..."
      return null
```
